src/RiskScore/RiskScoreConverter.ts

```diff
@@ -0,0 +1,230 @@
+            const {
+                externalId,
+                name,
+                effectiveDate,
+                riskScore,
+                alignmentScore,
+                rSquared,
+                retainedWeightProxied,
+                scoringMethodUsed
+            } = riskScorePortfolio.portfolio;
+
+            sortedPortfolios.push({
+                externalId,
+                name,
+                effectiveDate: Date.parse(effectiveDate),
+                riskScore,
+                alignmentScore,
+                rSquared,
+                retainedWeightProxied,
+                scoringMethodUsed
+            });
+
+            if (
+                riskScorePortfolio.metadata !== undefined && 
+                riskScorePortfolio.metadata.messages.length > 0
+            ) {
+                for (const message of riskScorePortfolio.metadata.messages) {
+                    const holdingNames = message.invalidHoldings.map(
+                        invalidHolding => invalidHolding.identifier
+                    );
+
+                    errors.push(
+                        `The holding(s) ${holdingNames.join(', ')} are invalid. ${message.message}`
+                    );
+                }
+            }
+        }
+
+        sortedPortfolios.sort((a, b) => (
+            a.effectiveDate === b.effectiveDate ?
+                0 :
+                a.effectiveDate < b.effectiveDate ? -1 : 1
+        ));
+
+        const portfolioNames = sortedPortfolios.map(portfolio => portfolio.name);
+
+        // Reset table
+
+        table.deleteColumns();
+
+        const valueColumns = [
+            'EffectiveDate',
+            'RiskScore',
+            'AlignmentScore',
+            'RSquared',
+            'RetainedWeightProxied',
+            'ScoringMethodUsed'
+        ];
+
+        for (const portfolioName of portfolioNames) {
+            for (const columnName of valueColumns) {
+                table.setColumn(`${portfolioName}_${columnName}`);
+            }
+        }
+
+        // Add risk scores to table
+
+        for (let rowIndex = 0; rowIndex < sortedPortfolios.length; rowIndex++) {
+            const portfolio = sortedPortfolios[rowIndex];
+            const riskScoreValues = [
+                portfolio.effectiveDate,
+                portfolio.riskScore,
+                portfolio.alignmentScore,
+                portfolio.rSquared,
+                portfolio.retainedWeightProxied,
+                portfolio.scoringMethodUsed
+            ];
+
+            for (let columnIndex = 0; columnIndex < riskScoreValues.length; columnIndex++) {
+                const column = `${portfolio.name}_${valueColumns[columnIndex]}`;
+                const value = riskScoreValues[columnIndex];
+                table.setCell(column, Number(rowIndex), value);
+            }
+        }
+
+        if (errors.length > 0) {
+            throw new Error(errors.join('\n'));
+        }
```

Comment:
Should this be handled differently?

This is the errors from invalid holdings. Morningstar still gives a result, and throwing the errors here will make the user unable to show the result.

Using console.error or equivalent makes the developer unable to handle the error in their own way.

Is there already a different way to signal errors that will not stop execution? If not, is it worth implementing?

Reply:
If the errors are part of the JSON, you can add them to a override of RiskScoreConnector.metadata. So you have to add an errors property to the converter to pass it on. Also you should stick to the original JSON error message.